File: J9Ch/src/J_4_Breadth_First_Search/Optional_6/_531_Six_Degrees_Medium.cpp
/** 531 Six Degrees */


/**
 * Definition for undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 */
class Solution {
public:
    /**
     * @param graph a list of Undirected graph node
     * @param s, t two Undirected graph nodes
     * @return an integer
     */
    int sixDegrees(vector<UndirectedGraphNode*> graph,
                   UndirectedGraphNode* s,
                   UndirectedGraphNode* t) {
        // Write your code here
        map<UndirectedGraphNode*, int> visited;
        queue<UndirectedGraphNode*> q;
        q.push(s);
        visited[s] = 0;

        while (!q.empty()) {
            UndirectedGraphNode* x = q.front();
            q.pop();
            if (x == t)
                break;

            for (UndirectedGraphNode* y : x->neighbors)
                if (visited.find(y) == visited.end()) {
                    visited[y] = visited[x] + 1;
                    q.push(y);
                }
        }

        if (visited.find(t) == visited.end())
            return -1;
        else
            return visited[t];
    }
};

File: J9Ch/src/J_4_BFS/Optional_6/python_cpp/_531_Six_Degrees_Medium.py
'''/** 531 Six Degrees '''


# Definition for Undirected graph node
# class UndirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []

import Queue

class Solution:
    '''
    @param {UndirectedGraphNode[]} graph a list of Undirected graph node
    @param {UndirectedGraphNode} s, t two Undirected graph nodes
    @return {int} an integer
    '''
    def sixDegrees(self, graph, s, t):
        # Write your code here
        d = {}
        q = Queue.Queue(maxsize = len(graph))

        q.put(s)
        d[s] = 0
        while not q.empty():
            x = q.get()
            if x == t:
                return d[x]

            for y in x.neighbors:
                if y not in d:
                    d[y] = d[x] + 1
                    q.put(y)

        return -1